layouts/BlogCategoryLayout.tsx

```diff
@@ -7,46 +7,58 @@ import { Time } from '@/components/Common/Time';
 import Link from '@/components/Link';
 import Pagination from '@/components/Pagination';
 import getBlogData from '@/next-data/blogData';
+import type { BlogDataRSC } from '@/types';
 
-const getCurrentCategory = (pathname: string) => {
+const getCategoryData = async (pathname: string) => {
   // We split the pathname to retrieve the blog category from it since the
   // URL is usually /blog/{category} the second path piece is usually the
   // category name, which usually year-YYYY
   const [, _pathname, category] = pathname.split('/');
-
   if (_pathname === 'blog' && category && category.length) {
-    return category;
+    const data = await getBlogData(category);
+    return {
+      ...data,
+      category,
+    };
   }
 
-  // if either the pathname does not match to a blog page
-  // which should not happen (as this hook should only be used in blog pages)
-  // or if there is no category in the URL we return the previous year as category name
-  // which is always the default category (for example, the blog index)
-  // see https://github.com/nodejs/nodejs.org/issues/6205 for a permanent solution
-  return `year-${new Date().getFullYear() - 1}`;
+  // If the pathname does not match to a blog page,
+  //   which should not happen (as this hook should only be used in blog pages),
+  // or, if there is no category in the URL,
+  //   which happens when we're on the blog overview page (index),
+  // then we return the most recent year with blog posts
+  let year = new Date().getFullYear() + 1;
+  let data: BlogDataRSC | undefined;
+  while (!data || (data.posts.length === 0 && year)) {
```

Comment:
Fwiw, I'm not completely a fan of this, but it keeps TS happy for implicit typing. `!data` is needed so `data` is implicitly narrowed to `BlogDataRSC` for the return, and it must be a standalone condition for that. I've included `year` as part of the condition as well, so that if there are no blog posts at all, we stop at year `0`, rather than going to `-Infinity`.